components/camel-kafka/src/main/java/org/apache/camel/processor/idempotent/kafka/KafkaIdempotentRepository.java

```diff
@@ -0,0 +1,342 @@
+        this.cache = Collections.synchronizedMap(new LRUCache<>(maxCacheSize));
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        log.info("Context: {}", camelContext);
+
+        // each consumer instance must have control over its own offset, so assign a groupID at random
+        String groupId = UUID.randomUUID().toString();
+        log.debug("Creating consumer with {}[{}]", ConsumerConfig.GROUP_ID_CONFIG, groupId);
+
+        consumerConfig.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
+        consumerConfig.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, Boolean.TRUE.toString());
+        consumerConfig.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+        consumerConfig.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+
+        consumer = new KafkaConsumer<>(consumerConfig);
+
+        producerConfig.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        producerConfig.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        // set up the producer to remove all batching on send
+        producerConfig.putIfAbsent(ProducerConfig.ACKS_CONFIG, "1");
+        producerConfig.putIfAbsent(ProducerConfig.BATCH_SIZE_CONFIG, "0");
+        producer = new KafkaProducer<>(producerConfig);
+
+        cacheReadyLatch = new CountDownLatch(1);
+        topicPoller = new TopicPoller(consumer, cacheReadyLatch);
+    }
+
+    @Override
+    public void setCamelContext(CamelContext camelContext) {
+        // doStart() has already been called at this point
+        this.camelContext = camelContext;
+        ExecutorServiceManager executorServiceManager = camelContext.getExecutorServiceManager();
+        executorService = executorServiceManager.newFixedThreadPool(this, "KafkaIdempotentRepository", 1);
+        executorService.submit(topicPoller);
+        log.info("Warming up cache");
+        try {
+            if (cacheReadyLatch.await(30, TimeUnit.SECONDS)) {
+                log.info("Cache OK");
+            } else {
+                log.warn("Timeout waiting for cache warm-up from topic {}. Proceeding anyway. "
+                        + "Duplicate records may not be detected.", topic);
+            }
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+    }
+
+    @Override
+    public CamelContext getCamelContext() {
+        return this.camelContext;
+    }
+
+    @Override
+    protected void doStop() {
+        // stop the thread
+        topicPoller.setRunning(false);
+        try {
+            if (topicPoller.getShutdownLatch().await(30, TimeUnit.SECONDS)) {
+                log.info("Expired waiting on topicPoller to shut down");
+            }
+        } catch (InterruptedException e) {
+            log.info("Interrupted waiting on latch: {}", e.getMessage());
+        }
+        executorService.shutdown();
```

Comment:
use camel context -> executor service manager -> has api to shutdown thread pool more safely